### PR TITLE
Enable Travis build for PowerPC-64LE build

### PR DESCRIPTION
**Pull request by kamasubb**
Added New Build Queue for PowerPC builds
Added changes to download and Install miniconda for PPC
Added changes to build conda-forge-build-setup

TODO:
- [x] Build su-exec locally and copy
- [x] Build tini locally and copy
- [x] Build conda-forge-ci-setup locally and copy
- [x] Merge https://github.com/conda-forge/conda-forge-pinning-feedstock/pull/78 and release
- [x] Merge https://github.com/conda-forge/docker-images/pull/67 and upload to dockerhub
- [x] Fix this PR to use docker.
- [ ] Create a help-ppc64le team
- [ ] Build su-exec using this PR
- [ ] Build tini using this PR
- [ ] Build conda-forge-ci-setup using this PR

### CODE REVIEW

**Comment by isuruf** [comment]
Here's how I'd like to see this implemented,
```yaml
# This file was generated automatically from conda-smithy. To update this configuration,
# update the conda-forge.yml and/or the recipe/meta.yaml.

language: generic

os: osx
osx_image: xcode6.4

env:
  global:
    # The BINSTAR_TOKEN secure variable. This is defined canonically in conda-forge.yml.
    - secure: ""

matrix:
  include:
    - env: CONFIG=osx_python2.7 MINICONDA_OS=MacOS-x86_64
      os: osx
      osx_image: xcode6.4
    - env: CONFIG=osx_python3.5 MINICONDA_OS=MacOS-x86_64
      os: osx
      osx_image: xcode6.4
    - env: CONFIG=osx_python3.6 MINICONDA_OS=MacOS-x86_64
      os: osx
      osx_image: xcode6.4

    - env: CONFIG=linux_ppc64le_python2.7 MINICONDA_OS=Linux-ppc64le
      os: linux_ppc64le
    - env: CONFIG=linux_ppc64le_python3.5 MINICONDA_OS=Linux-ppc64le
      os: linux_ppc64le
    - env: CONFIG=linux_ppc64le_python3.6 MINICONDA_OS=Linux-ppc64le
      os: linux_ppc64le


before_install:
    # Fast finish the PR.
    - |
      (curl https://raw.githubusercontent.com/conda-forge/conda-forge-ci-setup-feedstock/master/recipe/ff_ci_pr_build.py | \
          python - -v --ci "travis" "${TRAVIS_REPO_SLUG}" "${TRAVIS_BUILD_NUMBER}" "${TRAVIS_PULL_REQUEST}") || exit 1

    # Remove homebrew.
    - |
      if [ "$TRAVIS_OS_NAME" == "osx" ]; then
          echo ""
          echo "Removing homebrew from Travis CI to avoid conflicts."
          curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/uninstall > ~/uninstall_homebrew
          chmod +x ~/uninstall_homebrew
          ~/uninstall_homebrew -fq
          rm ~/uninstall_homebrew
      fi


install:
    # Install Miniconda.
    - |
      echo ""
      echo "Installing a fresh version of Miniconda."
      MINICONDA_URL="https://repo.continuum.io/miniconda"
      MINICONDA_FILE="Miniconda3-latest-${MINICONDA_OS}.sh"
      curl -L -O "${MINICONDA_URL}/${MINICONDA_FILE}"
      bash $MINICONDA_FILE -b

    # Configure conda.
    - |
      echo ""
      echo "Configuring conda."
      source ${HOME}/miniconda3/bin/activate root
      conda config --remove channels defaults
      conda config --add channels defaults
      conda config --add channels conda-forge
      conda config --set show_channel_urls true
      conda install --yes --quiet conda-forge-ci-setup=1
      source run_conda_forge_build_setup

script:
  - conda build ./recipe -m ./.ci_support/${CONFIG}.yaml

  - upload_or_check_non_existence ./recipe conda-forge --channel=main -m ./.ci_support/${CONFIG}.yaml
```

**Comment by isuruf** [comment]
With docker,

```yaml
# This file was generated automatically from conda-smithy. To update this configuration,
# update the conda-forge.yml and/or the recipe/meta.yaml.

language: generic

env:
  global:
    # The BINSTAR_TOKEN secure variable. This is defined canonically in conda-forge.yml.
    - secure: "ov4nUdga5WNh9X38+KqfK6cIjp7P5w6iPTsvW9tBYeiUZDKBelwZXjDtAs7AUNxLkqn8QucpowWhgIF838/iTynvTqbNOgoW8p07oqHdfbIP3of7aZ4zC2gbGPka9wiCj25mrEbSn931THi5bHrmxgqFMA9QBbdiIN0yzEIE46VGnz674aCmjeH6SeqKCPY6N0j8gvFZMWcb6EDhJHo5HnluluIR9V6sTJwlU5KRuUk+YTUaRg3brWjzQbLPE5WYZMdtwQNgAMyuYW1ly4vMy2f2acwJmfpSDlxMbarAhtO7QjZ3oPWMrbBf7qCeyAmp9chP078EPLXPAVDMUZkiAbOkLDOqAMQbV98rycFnN8/SbGlNhsUdaUAMjYbJOBiwH27JA39q8LPWH77MVQhQkSKHiOFXQoDFguCP8BI4Nz1wwf8G+bBxSzsnTRY2iCDSizMu3w+cHgWUTx7Mb/vVBHQMVnwriwInEWSA7QfIX1HAdTe3p22o5FZDKZFJK4uyoGKXM8Y9Ld6YpJ4LvT+/AZOxoEDt4mye//UOMRsQBx7/KE/l8zMuIFvDWLFTUZVJJaB5d2wvPrYl4B7beh1Js2L2GHgtI8b4Z6zEp3j1lvzhpwilvvIJ/T6aBIR8pzvWuMffAj4dfCsF+tvcinT0Nt4IZM+vDNqsb+DIyGfax7w="

matrix:
  include:
    - env: CONFIG=osx_python2.7
      os: osx
      osx_image: xcode6.4
    - env: CONFIG=osx_python3.5
      os: osx
      osx_image: xcode6.4
    - env: CONFIG=osx_python3.6
      os: osx
      osx_image: xcode6.4

    - env: CONFIG=linux_ppc64le_python2.7
      os: linux-ppc64le
    - env: CONFIG=linux_ppc64le_python3.5
      os: linux-ppc64le
    - env: CONFIG=linux_ppc64le_python3.6
      os: linux-ppc64le


before_install:
    # Fast finish the PR.
    - |
      (curl https://raw.githubusercontent.com/conda-forge/conda-forge-ci-setup-feedstock/master/recipe/ff_ci_pr_build.py | \
          python - -v --ci "travis" "${TRAVIS_REPO_SLUG}" "${TRAVIS_BUILD_NUMBER}" "${TRAVIS_PULL_REQUEST}") || exit 1

    # Remove homebrew.
    - |
      if [ "$TRAVIS_OS_NAME" == "osx" ]; then
          echo ""
          echo "Removing homebrew from Travis CI to avoid conflicts."
          curl -fsSL https://raw.githubusercontent.com/Homebrew/install/master/uninstall > ~/uninstall_homebrew
          chmod +x ~/uninstall_homebrew
          ~/uninstall_homebrew -fq
          rm ~/uninstall_homebrew
      fi


install:
    # Install Miniconda.
    - |
      if [ "$TRAVIS_OS_NAME" == "osx" ]; then
          echo ""
          echo "Installing a fresh version of Miniconda."
          MINICONDA_URL="https://repo.continuum.io/miniconda"
          MINICONDA_FILE="Miniconda3-latest-MacOSX-x86_64.sh"
          curl -L -O "${MINICONDA_URL}/${MINICONDA_FILE}"
          bash $MINICONDA_FILE -b
      fi

    # Configure conda.
    - |
      if [ "$TRAVIS_OS_NAME" == "osx" ]; then
          echo ""
          echo "Configuring conda."
          source ${HOME}/miniconda3/bin/activate root
          conda config --remove channels defaults
          conda config --add channels defaults
          conda config --add channels conda-forge
          conda config --set show_channel_urls true
          conda install --yes --quiet conda-forge-ci-setup=1
          source run_conda_forge_build_setup
      fi

    - |
      if [ "$TRAVIS_OS_NAME" != "osx" ]; then
          docker pull kamasubb/conda-forge-linux-anvil-ppc64le-centos
      fi

script:
    - |
      if [ "$TRAVIS_OS_NAME" == "osx" ]; then
          conda build ./recipe -m ./.ci_support/${CONFIG}.yaml;
          upload_or_check_non_existence ./recipe conda-forge --channel=main -m ./.ci_support/${CONFIG}.yaml
      fi

    - |
      if [ "$TRAVIS_OS_NAME" != "osx" ]; then
          ./.circleci/run_docker_build.sh
      fi
```

**Comment by scopatz** [comment]
Is this really needed?  We dropped Windows 32 bit.  Maybe we could have a discussion about this on the next dev call?

**Comment by ocefpaf** [comment]
If it is on a separate queue Travis-CI and the maintenance burden won't fall on the core members shoulder I'm OK with it.

**Comment by scopatz** [comment]
Yeah, that is fair.  I just worry that there will potentially be a lot a recipes where PPC will break and we'll end up needing to skip this host. Maybe it won't be so bad, though.

**Comment by ocefpaf** [comment]
> I just worry that there will potentially be a lot a recipes where PPC will break and we'll end up needing to skip this host.

Yep. That may be a burden...

**Comment by isuruf** [comment]
I think there are 2 issues.

1. Adding support in conda-smithy
    The platform can be disabled by default. win-32 is still supported by conda-smithy with a different `conda_build_config.yaml`
 
2. Enabling ppc64le in feedstocks.

**Comment by scopatz** [comment]
I am totally happy if this ends up being opt-in :smile:

**Comment by jakirkham** [comment]
To make sure no one missed out, cc @conda-forge/core 😉

**Comment by jayfurmanek** [comment]
Hi all, 
Some anecdotal feedback - most of the conda recipes I've tried on ppc64le have built just fine. I can help maintain any failures.

@isuruf 
What else is needed to get this PR in?

Thanks!

**Comment by isuruf** [comment]
@jayfurmanek, first we need to get a docker image ready. Since anaconda seems to be using centos7 for ppc64le, let's go with a centos7 based docker image. Can you send a PR to https://github.com/conda-forge/docker-images with a DockerFile for ppc64le?

**Comment by jjhelmus** [comment]
CentOS 7 was the first version of CentOS that included support for the ppc64le architecture.  Similarity Ubuntu 14.04 was the first release for Ubuntu.  CentOS 7 uses an earlier glibc (2.17) than Ubuntu (2.19) which is why it was choose as the base image.

**Comment by jayfurmanek** [comment]
Sure thing. I think @kamasubb has one ready. I'll work with him to get a PR up.

**Comment by jakirkham** [comment]
Are there things that we need to do to existing feedstocks (beyond re-rendering) to get this to work?

**Comment by CJ-Wright** [comment]
@jakirkham is this similar to the new compilers where we need to build in order?

**Comment by jjhelmus** [comment]
The [docker image from the conda-concourse-ci repository](https://github.com/conda/conda-concourse-ci/tree/master/docker/c3i-linux-ppc64le) is used to build the packages in `defaults`.  If conda-forge used that image as well, or one with a few modification, compatibility with `defaults` would be excellent.

**Comment by jayfurmanek** [comment]
@isuruf @jjhelmus I've prepared two different images for discussion. The first is similar to the existing image at the docker_images/linux-anvil:
https://github.com/jayfurmanek/docker-images/tree/ppc64le

The second is closer to `defaults`, which is simpler.
https://github.com/jayfurmanek/docker-images/tree/ppc64le-defaults

The first image uses an entrypoint to set up some users and persmissions, etc. The defaults image does not. I don't know for sure if that set up is needed. I've tested both as far as doing the docker build and manually starting a container.

Is there a preference in which branch I should raise the PR for?

**Comment by isuruf** [comment]
I think we need the one based on linux-anvil, but maybe add more of the yum packages from defaults. @jjhelmus, what do you think?

**Comment by jayfurmanek** [comment]
The main package differences looks to be the inclusion of more *-devel packages in the defaults image. I can certainly add those into the anvil based one.

**Comment by jjhelmus** [comment]
> @jjhelmus, what do you think?
This decision needs to be made by the wider conda-forge community, I should not be responsible for this.

**Comment by isuruf** [comment]
@conda-forge/core, what do you think?

**Comment by isuruf** [comment]
@jayfurmanek, can you make a PR with linux-anvil based image with the extra dependencies from the defaults image? We can discuss more in the PR

**Comment by jayfurmanek** [comment]
Yes! Sounds like a plan. Thanks!

**Comment by jayfurmanek** [comment]
https://github.com/conda-forge/docker-images/pull/67

**Comment by ipoddar-ibm** [comment]
@isuruf and @jakirkham now that we have a ppc64le linux-anvil image, can we make some progress on this PR? thanks IP

File: conda_smithy/configure_feedstock.py [hunk 1]
```diff
@@ -668,15 +674,17 @@ def copy_feedstock_content(forge_dir):
 def _load_forge_config(forge_dir, exclusive_config_file):
     config = {'docker': {'executable': 'docker',
                          'image': 'condaforge/linux-anvil',
+                         'image_ppc64le': 'condaforge/linux-ppc64le-anvil',
```
Comment (isuruf) [inline conda_smithy/configure_feedstock.py]: Needs to be `condaforge/linux-anvil2`

**Comment by ipoddar-ibm** [comment]
@isuruf any thoughts on how to address the failing check for coverage? can this PR be merged?

**Comment by jayfurmanek** [comment]
There is an open bug for the failing test here. It's unrelated to this PR:
https://github.com/conda-forge/conda-smithy/issues/844

**Comment by ipoddar-ibm** [comment]
@jakirkham are the conda-forge builds now running on Travis CI on POWER? Are there any recipes that are getting built on POWER and published?

**Comment by CJ-Wright** [comment]
Would it be possible to rebase this onto master?

**Comment by jayfurmanek** [comment]
I attempted a rebase with pull 882 back in August (referenced above), but didn't finish it. A this point, it'll probably need a fresh look.
@kamasubb do you have time to look at this?

**Comment by isuruf** [comment]
Done in #1023

**Comment by ipoddar-ibm** [comment]
@isuruf thanks  a  lot for  getting ppc64le builds working.  Is there one or more  travis-ci.org URLs where we can look at the status of  these builds?

**Comment by isuruf** [comment]
There's https://travis-ci.org/conda-forge/, but it's hard to only see ppc64le builds.
Packages already built can be found at https://conda.anaconda.org/conda-forge/linux-ppc64le/

**Comment by mariusvniekerk** [comment]
You can also track the ppc64le migration at https://conda-forge.org/status/

**Comment by ipoddar-ibm** [comment]
@mariusvniekerk thanks, I see that there are 20 awaiting-parents, are you expecting the community to submit pull requests for these?

**Pull request closed without merging**